src/validation.cpp

```diff
@@ -2185,16 +2189,46 @@ void EnsureTxData(const CTransaction& tx, const CCoinsViewCache& inputs, Precomp
     assert(txdata.m_spent_outputs.size() == tx.vin.size());
 }
 
+/**
+ * Turn all of this transaction's input scripts into script checks and return them as a vector.
+ * Any script checks which are not necessary (eg due to script execution cache hits) are excluded.
+ *
+ * Setting cacheSigStore/cacheFullScriptStore to false will remove elements from the corresponding cache
+ * which are matched. This is useful for checking blocks where we will likely never need the cache
+ * entry again.
+ *
+ * Non-static (and redeclared) in src/test/txvalidationcache_tests.cpp
+ */
+std::vector<CScriptCheck> PrepareInputScriptChecks(const CTransaction& tx,
+                       const CCoinsViewCache& inputs, unsigned int flags, bool cacheSigStore,
+                       bool cacheFullScriptStore, PrecomputedTransactionData& txdata,
+                       ValidationCache& validation_cache) EXCLUSIVE_LOCKS_REQUIRED(cs_main)
+{
+    if (tx.IsCoinBase()) return {};
+
+    uint256 hashCacheEntry = GetHashCacheEntry(validation_cache, tx, flags);
+    if (IsScriptCached(validation_cache, cacheFullScriptStore, hashCacheEntry)) {
+        return {};
+    }
+
+    EnsureTxData(tx, inputs, txdata);
+
+    std::vector<CScriptCheck> pvChecks{};
```

Comment:
Just a rough idea, but did you consider the alternative approach of doing the `pvChecks` population not here (but in `ConnectBlock()` or a separate function), so that `CheckInputScripts` could also call `PrepareInputScriptChecks` and there would be no duplication of the coinbase check, cache check and `TxData` init?